googlecloudrobotics/core
# Add certificate authority selection support and Google CAS integration

Closes #107.

This PR introduces the concept of Certificate Authority (CA) providers to deployment scripts. It allows selecting a certificate provider to configure Terraform files, helm charts and cluster resources necessary to generate certificates using the selected provider. 

The existing integration with Let's Encrypt has been kept as a certificate provider option. Integration with Google's Certificate Authority Service (CAS) has been added and can be selected as a certificate provider.

## Changes

1. Certificate providers. There are two options available: `lets-encrypt`, and `google-cas`. The default option was set to `lets-encrypt` for backwards compatibility purposes.
2. Some configuration variables have been added. Existing configurations can be updated by running `deploy.sh set-config`.

## How to test

1. Run `./deploy.sh set-config <PROJECT_ID>` to configure new environment variables.
    * Select `google-cas` as the certificate provider.
    * Fill in the Common Name (`CN`) and Organization (`O`) certificate fields. Optionally, fill in the Organizational Unit (`OU`) certificate field.
2. Run `./deploy.sh update <PROJECT_ID>` to deploy the configuration.
3. Verify that the `cloud-robotics` certificate was issued correctly.

The output should look something like this

```bash
# Make sure kubectl is using the correct config and context
$ kubectl get certificate cloud-robotics 
NAME             READY   SECRET               AGE
cloud-robotics   True    cloud-robotics-tls   XXm
```

4. Verify that `www.endpoints.robco-ensonic.cloud.goog` is using the certificate.

```sh
# Make sure to set <PROJECT_ID>
openssl s_client -showcerts -connect www.endpoints.$PROJECT_ID.cloud.goog:443
```
